### DISCUSSION TITLE
Suspend and resume activity in loop

### DISCUSSION THREAD

**Discussion by msciancalepore** [elsa-workflows/elsa-core#2599] [Q&A]
Hello guys,
thank you for the grat work you are doing.
I need some help usign elsa workflow, in particular I would like to suspend and resume a custom activity in a loop, that is suspend and resume more times.
I followed the guide https://elsa-workflows.github.io/elsa-core/docs/guides/guides-blocking-activities and everything works without loop. I'm able to suspend and resume an activity one time.
If I put the suspended/resumed activity in a loop, Rebus Exception will raise.

This is my workflow

![elsa_workflow_loop](https://user-images.githubusercontent.com/96257912/146423732-c11fc6de-75ea-42b8-bf33-a4437413337f.JPG)

and this is the custom activity that is suspended and resumed (ExecuteFirstCommand)
![elsa_custom_activity](https://user-images.githubusercontent.com/96257912/146423861-dc83408e-3670-4d93-8d9e-79cb2d2fd39d.JPG)

I have some dependency injections and in the onExecute method I invoke some services that will resume the activity in asynchronous way (through bookmarks and invokers, like is described in the guide).
This is the exception that is raised
![rebus_exception](https://user-images.githubusercontent.com/96257912/146424329-c475b009-336b-49cc-a5b8-ae981c62c2a5.JPG)

I resume the activity with

**await _workflowLaunchpad.CollectAndDispatchWorkflowsAsync(context, null, cancellationToken);**

I don't know if this is the right way to use Elsa workflow, any suggestions will be more appreciate.
Thank you.

EDIT:
I've tried with parent/child workflows, following the guide https://elsa-workflows.github.io/elsa-core/docs/guides/guides-parent-child-workflows
These are my workflows
![readorder](https://user-images.githubusercontent.com/96257912/146446005-8b33c6db-509c-4daa-b31d-8cb6f38b72c2.PNG)
![sendorder](https://user-images.githubusercontent.com/96257912/146446016-b3eb73a9-c83d-4c2a-b3c9-1c966713e1c1.JPG)

and I used the Isignaler interface to send signal "CommandExecuted".
 **await _signaler.DispatchSignalAsync(signal: "CommandExecuted", workflowInstanceId: _contextId);**
Same situation, first time it works, second time Rebus exception
**Could not 'GetOrAdd' item with key 'outgoing-messages' as type System.Collections.Concurrent.ConcurrentQueue`1[Rebus.Transport.AbstractRebusTransport+OutgoingMessage]**

**Comment by sfmskywalker** [comment]
Any chance you can share your (stripped down) project so I can take a look?

**Comment by msciancalepore** [comment]
yes, I can, Thank you.
These are my elsa json workflows
[elsaJson.zip](https://github.com/elsa-workflows/elsa-core/files/7746211/elsaJson.zip)
And these are the classes: 
[OrderHandler.zip](https://github.com/elsa-workflows/elsa-core/files/7746224/OrderHandler.zip)
ExecuteFirstCommand (in the Custom Activities folder) is the activity that I'm trying to suspend and resume.
I suspend it using the service **resumeExecuteFirstCommand** in the OnExecute() method, which subscribes to an external asynchronus event. The same service should resume the activity.
I found a workaround and I'm using it at moment: calling a http post, instead of a singaler or invoker call, is working.
You can see in the ResumeCommandExecutedAsync() method of the **ResumeExecuteFirstCommand** service (in the folder Services) the commented code I was trying to use.
If it's not enough to understand, I can explain it more in details.
Thank you.

**Reply by sfmskywalker** [reply]
Hi @msciancalepore , my apologies for the very late response. Not sure if you ever figured it out, but if you still need help, please send me a project (including Program.cs etc) - this will make it easier for me to run the project instead of trying to reconstruct using individual files.
